colink/p2p_inbox.py

```diff
@@ -0,0 +1,226 @@
+from typing import Mapping, Set
+import colink as CL
+import jwt
+import json
+import secrets
+from http.client import HTTPSConnection
+from http.server import HTTPServer, BaseHTTPRequestHandler
+import ssl
+import socket
+import random
+import threading
+from queue import Queue
+import time
+from cryptography import x509
+from cryptography.hazmat.primitives.serialization import Encoding
+from tempfile import NamedTemporaryFile
+from .tls_utils import gen_cert
+
+
+class VTInbox:
+    def __init__(self, addr: str, vt_jwt: str, tls_cert: bytes):
+        self.addr = addr
+        self.vt_jwt = vt_jwt
+        self.tls_cert = tls_cert
+
+
+Status_OK = 200
+Status_BAD_REQUEST = 400
+Status_UNAUTHORIZED = 401
+
+
+class VTInBox_RequestHandler(BaseHTTPRequestHandler):
+    def _send_response(self, resp: int):
+        self.send_response_only(resp)
+        self.send_header("Content-type", "text/plaintext")
+        self.end_headers()
+
+    def do_POST(self):
+        data = self.server.data
+        notification_channels = self.server.notification_channels
+        user_id = self.headers.get("user_id", "")
+        key = self.headers.get("key", "")
+        token = self.headers.get("token")
+        if not user_id or not key or not token:
+            self._send_response(Status_BAD_REQUEST)
+            return
+        try:
+            token = jwt.decode(token, self.server.jwt_secret, algorithms=["HS256"])
+        except Exception as e:
+            self._send_response(Status_UNAUTHORIZED)
+            return
+
+        if token["user_id"] != user_id:
+            self._send_response(Status_UNAUTHORIZED)
+            return
+        # payload
+        length = int(self.headers.get("content-length"))
+        body = self.rfile.read(length)
+        data[(user_id, key)] = body
+        nc = notification_channels.get((user_id, key), "")
+        if nc:
+            nc.put(1)  # send
+        self._send_response(Status_OK)
+
+
+def server_maintainer(server, q):
+    while True:
+        if not q.empty():
```

Comment:
I think instead of using a separate queue, `threading.Condition` might look clearer